### PR TITLE
Canvas rotation

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request uses css canvas rotation to allow for on-the-fly image rotation (original effort: #1384).

[Plugin docs](https://jdaviz--1983.org.readthedocs.build/en/1983/imviz/plugins.html#canvas-rotation)
[Plugin API docs](https://jdaviz--1983.org.readthedocs.build/en/1983/api/jdaviz.configs.imviz.plugins.rotate_canvas.rotate_canvas.RotateCanvas.html#jdaviz.configs.imviz.plugins.rotate_canvas.rotate_canvas.RotateCanvas)

https://user-images.githubusercontent.com/877591/221963634-e9610b4d-cf9e-4db8-b431-81c54569d769.mov

**TODO**:
- [x] cross-browser support for the correct dragging direction.  Will need significant testing across different browsers/versions and perhaps considering workarounds upstream in glue/bqplot or worse case scenario we could detect compatibility and display a message in the plugin
- [x] cleanup plugin UI (live update instead of button press?  option to set north to up? rename plugin to distinguis from reproject)
- [x] Have zoom box in Compass follow canvas rotation. (Disabled for now.). Or for canvas itself to rotate?
- [x] Multi-viewer support (value should update when viewer dropdown is changed, consider support for multiselect a la plot options... what to do with linked zoom limits in Imviz...)
- [x] Port some of the relevant docs from #1340 into `docs/imviz/plugins.rst`. This should wait till after #1393 is merged to avoid conflicts.
- [x] Test coverage?
- [x] Any other TODOs in the diff

**Follow-up**:
- [ ] track down and fix the 1/3 image disappearing bug.  Probably was introduced somewhere in the limits logic since it does not seem to happen on main. 
- [ ] Home button no longer make sense when canvas is rotated. Is this something we can address?  Should rotation reset or just set limits such that the whole image is in-frame?
- [ ] Re-consider canvas oversizing (to avoid cropping on the edge of the figures by extending beyond the viewer but faking the zoom-level... see several different attempts at doing that in the commit history in this PR).
- [ ] this currently works per-viewer, but could be changed to be global or the viewer dropdown could support multi-select (with logic for mixed state either copied from plot options or generalized or a lock button that just adopts the default viewer's angle for all viewers and hides the dropdown)

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by kecnry** [comment]
Note that https://github.com/bqplot/bqplot/pull/1585 is required to fix the drag direction in Firefox and Safari.  Do we want to hold this up until that is merged/released/pinned?

**Comment by rosteen** [comment]
> Note that [bqplot/bqplot#1585](https://github.com/bqplot/bqplot/pull/1585) is required to fix the drag direction in Firefox and Safari. Do we want to hold this up until that is merged/released/pinned?

I consider it a blocker, personally, the behavior is really annoying.

**Comment by kecnry** [comment]
> I consider it a blocker, personally, the behavior is really annoying.

Ok, I think this can still be reviewed (on other browsers or using that PR), but we can defer merging until that can be pinned.

**Comment by camipacifici** [comment]
In Chrome, everything works as expected using a single viewer. I see in the task list to work on the zoom box in compass. I do not see it as a blocker, but it will be a nice to have as a follow up.
When I load a second viewer, I can rotate using the slider, but the present selections do nothing (they still work if I select the first viewer).
About the home button: to me it makes sense that it resets the zoom, but preserves the rotation. Happy to discuss other options.
With the two viewers, if I zoom (with the linked option) on the first viewer, the second viewer "blinks" (for lack of a better word, although it is not an actual blinking between two different images) a couple of times before settling on the correct zoom.
Screen Shot 2023-03-02 at 12 47 14 PM

**Comment by kecnry** [comment]
> In Chrome, everything works as expected using a single viewer. I see in the task list to work on the zoom box in compass. I do not see it as a blocker, but it will be a nice to have as a follow up.

I actually think the lack of zoom box might be because of distortion (@pllim is that right)?  This PR now rotates the whole compass image, so if @pllim can confirm that, I'll remove it from the follow-up list.

> When I load a second viewer, I can rotate using the slider, but the present selections do nothing (they still work if I select the first viewer).

Fixed!

> About the home button: to me it makes sense that it resets the zoom, but preserves the rotation. Happy to discuss other options.

I agree, I think the home button discussion is coupled to the possibility of "oversizing/extending" the canvas underneath to avoid the cropped corners.  I removed that from this PR for now since I'm not sure we want it and it seemed unreliable - but is something we might want to revisit.

> With the two viewers, if I zoom (with the linked option) on the first viewer, the second viewer "blinks" (for lack of a better word, although it is not an actual blinking between two different images) a couple of times before settling on the correct zoom.

I can't reproduce this beyond the normal resolution adjustment.  Do you see the same when unrotated?  I'm guessing that is something we should look into upstream, but not 100% sure what would be causing that.

**Comment by camipacifici** [comment]
> > In Chrome, everything works as expected using a single viewer. I see in the task list to work on the zoom box in compass. I do not see it as a blocker, but it will be a nice to have as a follow up.
> 
> I actually think the lack of zoom box might be because of distortion (@pllim is that right)? This PR now rotates the whole compass image, so if @pllim can confirm that, I'll remove it from the follow-up list.
> 
> > When I load a second viewer, I can rotate using the slider, but the present selections do nothing (they still work if I select the first viewer).
> 
> Fixed!
> 
> > About the home button: to me it makes sense that it resets the zoom, but preserves the rotation. Happy to discuss other options.
> 
> I agree, I think the home button discussion is coupled to the possibility of "oversizing/extending" the canvas underneath to avoid the cropped corners. I removed that from this PR for now since I'm not sure we want it and it seemed unreliable - but is something we might want to revisit.
> 
> > With the two viewers, if I zoom (with the linked option) on the first viewer, the second viewer "blinks" (for lack of a better word, although it is not an actual blinking between two different images) a couple of times before settling on the correct zoom.
> 
> I can't reproduce this beyond the normal resolution adjustment. Do you see the same when unrotated? I'm guessing that is something we should look into upstream, but not 100% sure what would be causing that.

I do see it regardless of rotation.

**Comment by pllim** [comment]
> the lack of zoom box might be because of distortion

I don't think so. The zoom box only cares about the 4 corners. If your data is distorted, the box can be misleading but should still show, theoretically.

File: jdaviz/app.py
Comment (pllim) [inline jdaviz/app.py:1649]: I feel this key name should be more specific in case someone asks for vertical flip in the future.
```suggestion
            'flip_horizontal': False,  # canvas rotation horizontal flip
```

File: jdaviz/configs/imviz/plugins/compass/compass.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/compass/compass.py:30]: Can we rename "canvas_rotatey" to something less cryptic? I have no idea what it is supposed to do by just the name without having to read the code. Is it just me?

File: jdaviz/configs/imviz/plugins/rotate_canvas/rotate_canvas.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/rotate_canvas/rotate_canvas.py:13]: `viewer_requirements='image'` eh? Are you sure users won't want to rotate spectrum viewer in the future so they don't have to subtract a monotonously increasing/decreasing continuum? 😉

File: jdaviz/configs/imviz/plugins/rotate_canvas/rotate_canvas.vue
Comment (pllim) [inline jdaviz/configs/imviz/plugins/rotate_canvas/rotate_canvas.vue:58]:
```suggestion
```

File: jdaviz/configs/imviz/plugins/viewers.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/viewers.py:237]: Ah, so you did end up disabling the zoom box in Compass? Should make a note in the plugin user doc that this will not show if canvas is rotated.

**Review by pllim — COMMENTED** [review]
(This is the general code review portion. I will play with the implementation next week and post more review comments, if applicable. Thanks for your patience!)

Can you move the `test_rotate_canvas.py` up a few levels to `imviz/tests` folder? I purposely not have `tests` folders in random `imviz` subfolders so I can easily track all the tests of Imviz in one place. 😅  (Ships had sailed for the other vizes by the time I joined the project...)

Is it too much to ask for an additional test for multi-viewer use case? 😅

File: jdaviz/configs/imviz/plugins/compass/compass.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/compass/compass.py:30]: I was a bit torn on that... `rotatey` is the css property that is being set, and this isn't intended to be exposed to the user.  I guess I could use the same `flip`/`flip_horizontal` boolean, but then do need logic on the JS side to map that to the CSS property, and I was trying to avoid duplicating that logic.  So... any preference between some duplicated JS code but consistent traitlets, or this somewhat cryptic traitlet name and adding an inline comment?

File: jdaviz/configs/imviz/plugins/viewers.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/viewers.py:237]: Hmmm.... I think you added this code.  Do you think now that we just rotate the compass image entirely that this can safely be removed?

File: jdaviz/configs/imviz/plugins/compass/compass.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/compass/compass.py]: I switched this to consistently use `flip_horizontal` (or `canvas_flip_horizontal` when stored outside the plugin).

File: jdaviz/configs/imviz/plugins/viewers.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/viewers.py]: I removed this code and clarified the behavior in the docs for both plugins.  Let me know what you think!

Comment (pllim) [reply jdaviz/configs/imviz/plugins/viewers.py]: It works better than I thought it would! 😸

File: jdaviz/configs/imviz/plugins/compass/compass.vue
Comment (pllim) [inline jdaviz/configs/imviz/plugins/compass/compass.vue:23]: 🤯

**Review by pllim — COMMENTED** [review]
Maybe I missed the conversation but what is the point of the switch to allow people to flip horizontally? A button to reset to no rotation would be more useful for me.

Not a deal breaker but I just want to note that I see white dotted lines and horizontal scrollbar on the display when rotated. Screenshot below.

Screenshot 2023-03-07 174540

Also not a deal breaker, but I do find the box zoom hard to use when rotated. I want to draw a horizontal box but it is slanted. I guess just a matter of getting used to the concept of rotated canvas.

After I box-zoomed into an area on rotated canvas, stuff outside the zoom box but technically still part of the image is now not "seen" by the display. Screenshot below.

Screenshot 2023-03-07 175436

Another way to look at this.

Screenshot 2023-03-07 181037

When I have two viewers, one rotated and one isn't. Then I used the "linked pan/zoom" (without canvas rotation, it is supposed to pan/zoom to the same area on the sky), and things look a little weird. Is this expected behavior? Maybe we already discussed this before but I don't remember.

Screenshot 2023-03-07 181244

### Conclusion

This is simple, scalable, and allows people to rotate view to their hearts content. Display cropping is a known limitation and should not block merge. 

I will approve if POs are okay with the behaviors noted above.

**Comment by pllim** [comment]
Check this out in a notebook! Adapted from `imviz_load_3d_slices.ipynb`. 😼 

```python
import time

import numpy as np
from astropy.utils.data import download_file
from PIL import Image, ImageSequence
from skimage.color import rgb2gray

from jdaviz import Imviz

imviz3 = Imviz()

gif_file = download_file('https://media4.giphy.com/media/J5pnZ53pj4cmu30Rx5/giphy.gif', cache=True)

im = Image.open(gif_file)
i = 0

with imviz3.batch_load():
    for frame in ImageSequence.Iterator(im):
        if i != 30:  # Skip some
            i += 1
            continue
        data = np.asarray(frame)[::-1, :]
        if data.ndim == 3:
            data = rgb2gray(data)
        imviz3.load_data(data, data_label=f'frame_{i}', do_link=False)
        i += 1

imviz3.show()

plg = imviz3.plugins["Canvas Rotation"]

while True:
    plg.angle += 1
    if plg.angle >= 360:
        plg.angle = 0
    time.sleep(0.1)
```

**Comment by kecnry** [comment]
> Maybe I missed the conversation but what is the point of the switch to allow people to flip horizontally? A button to reset to no rotation would be more useful for me.

The flip allows for always supporting north-up east-left (and east-right) presets.  I can easily add a reset button (now or later) if that would be a useful feature.

> Not a deal breaker but I just want to note that I see white dotted lines and horizontal scrollbar on the display when rotated.

I do not see these, but will try to hide the scrollbar (not sure there is anything I can do about the lines).

> Also not a deal breaker, but I do find the box zoom hard to use when rotated. I want to draw a horizontal box but it is slanted. I guess just a matter of getting used to the concept of rotated canvas.

Yes, that is just how this method of rotation works.  If you think of the pixels themselves, it makes sense that subsets and zoom limits need to follow the direction of the pixels (which have not been reprojected).

> After I box-zoomed into an area on rotated canvas, stuff outside the zoom box but technically still part of the image is now not "seen" by the display. Screenshot below.

Right, this was the "oversizing" effort that was abandoned in order to get this in, but could be revisited in the future.  In order to not have this, we have to oversize the underlying canvas and then fake the zoom-limits on the axes so that the requested zoom limits fall within the viewer.  This then gets quite confusing from API calls/state for zoom-limits though...

**Comment by kecnry** [comment]
If the goal of this is to present the image in the orientation the user is used to, that in theory could be some non-north-up-east-left but still flipped from the pixels... so I think its useful to have as an option still.  Do you just find it confusing?  Is that something we could address in the text or docs instead of disabling?

**Comment by pllim** [comment]
> Do you just find it confusing?

Yes. Because if you want to make it that flexible and yet there is no flip vertical, it makes me question things.

**Comment by kecnry** [comment]
A rotation and single flip is enough to reach any orientation (a vertical flip is the same as a 180 deg rotation then a horizontal flip).  I do think having access to the flip is important to allow achieving any orientation.... and if we add vertical flip for the sake of completeness, then I worry order of operations just becomes increasingly confusing.

**Comment by pllim** [comment]
Re: scrollbar -- I cannot find the element that highlights the horizontal scrollbar. Is there a way to click on the scrollbar and have Elements jump to the, uh, element?

I see this though in dev console. Is it related?

Screenshot 2023-03-08 101547

**Comment by pllim** [comment]
> In chrome, you want the button in the top-left

I found the Elements tab. But I don't know how to comb through it. Maybe we'll need to Slack call or something when we have time.

File: jdaviz/configs/imviz/plugins/rotate_canvas/tests/test_rotate_canvas.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/rotate_canvas/tests/test_rotate_canvas.py:7]: Please move this test file up to `imviz/tests` folder. Thanks!

**Comment by rosteen** [comment]
Ori gave the thumbs-up to approve this in its current state after seeing a demo at the sprint review.

**Comment by pllim** [comment]
@kecnry , the horizontal scrollbar is still here even with https://github.com/spacetelescope/jdaviz/pull/2078 applied. But I don't see the JS error in console log anymore.

**Comment by pllim** [comment]
I found another bug, or is this a feature?

1. Fire up Imviz example notebook.
2. Load the data.
3. Link them by WCS.
4. Blink to "A".
5. Open second viewer.
6. Load "B" in second viewer.
7. Use linked pan/zoom to zoom/pan a little in one of them.
8. Open Canvas Rotation.
9. Apply N-up E-left to both viewers.
10. Use viewer drop down in Canvas Rotation to switch viewer.
11. See the viewer being switched to resets on its own. I would expect the viewer to "remember" the rotation but this is not the case.

**Comment by kecnry** [comment]
@pllim - fixed, thanks!

**Review by pllim — APPROVED** [review]
As it is, I think it is good enough to merge. Also, Ori is happy, so I am approving. Thanks!

File: jdaviz/configs/imviz/plugins/compass/compass.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/compass/compass.py:61]: Nitpick: Not sure why `# noqa` is needed here.

```suggestion
        self.canvas_angle = viewer_item.get('canvas_angle', 0)
```

**Review by rosteen — APPROVED** [review]
I confirmed that the panning direction problem is fixed with the most recent `bqplot`, approved.

**Pull request merged**
